## This notebook applies h2o auto ML techniques to predict Redshift through linear regression models.

In [35]:
%load_ext autoreload
%autoreload 2
import os 
import sys
import time

import pandas as pd

from sklearn.metrics import mean_squared_error

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [30]:
module_path = os.path.abspath(os.path.join('..','..'))
if module_path not in sys.path:
    sys.path.append(module_path)

from src.modeling.functions_autoML import auto_ML, save_aml_models, gen_predictions

In [3]:
pd.set_option("display.max_rows", 6)

In [21]:
df = pd.read_pickle("../../data/structured/df.pkl")
#IA Filter
df = df.loc[df.type_bool == True]
df

desg  \
ID                                                                                
SN10005   [[10.027000000001863, 4.996, 8.081], [33.945000000006985, 1.166, 3...   
SN10007   [[0.0, 6.266, 10.48], [10.105999999999767, 17.69, 25.58], [29.9260...   
SN100361  [[6.957000000002154, 16.02, 23.92], [7.969000000004598, 7.304, 6.5...   
...                                                                         ...   
SN99438   [[0.0, 2.33, 2.55], [7.894000000000233, 3.805, 5.684], [13.9919999...   
SN99457   [[0.0, 10.06, 4.768], [4.075000000004366, 4.395, 2.551], [11.96900...   
SN99946   [[11.961000000002969, 1.526, 5.683], [13.055000000000291, 0.4911, ...   

                                                                           desi  \
ID                                                                                
SN10005   [[0.015000000006693881, 0.2977, 2.206], [18.94100000000617, 4.11, ...   
SN10007   [[0.03200000000651926, 7.161, 9.313], [2.0160000000032596, 11.87, ...   
SN100361  [[0.023000000001047738, 4.903, 3.324], [6.984000000004016, 0.4745,...   
...                                                                         ...   
SN99438   [[0.030999999995401595, 0.006543, 3.683], [7.944999999999709, 1.66...   
SN99457   [[3.0429999999978463, 0.9564, 1.984], [4.105999999999767, 3.45, 3....   
SN99946   [[0.01599999999598367, 4.95, 3.147], [3.0349999999962165, 0.5629, ...   

                                                                           desr  \
ID                                                                                
SN10005   [[0.0, 1.033, 1.325], [13.015000000006694, 0.4544, 1.853], [33.969...   
SN10007   [[2.0, 1.24, 2.429], [10.121000000006461, 20.06, 4.181], [15.98900...   
SN100361  [[0.0, 0.7103, 2.626], [5.9760000000023865, 5.207, 2.392], [6.9690...   
...                                                                         ...   
SN99438   [[0.014999999999417923, 0.4446, 1.3], [14.003999999993539, 0.531, ...   
SN99457   [[3.004000000000815, 1.456, 1.709], [4.0900000000037835, 0.4515, 1...   
SN99946   [[0.0, 0.4826, 2.129], [2.995999999999185, 0.7092, 1.709], [4.0820...   

                                                                           desz  \
ID                                                                                
SN10005   [[0.03500000000349246, 1.826, 3.029], [48.875, 1.71, 2.119], [57.8...   
SN10007   [[0.9770000000062282, 0.6586, 5.329], [15.993000000002212, 29.8, 6...   
SN100361  [[8.055000000000291, 2.567, 2.665], [15.035000000003492, 3.775, 3....   
...                                                                         ...   
SN99438   [[0.06199999999807915, 1.497, 4.486], [7.971999999994296, 8.596, 3...   
SN99457   [[4.137000000002445, 9.726, 4.49], [12.046999999998661, 6.372, 3.4...   
SN99946   [[0.03899999999703141, 1.045, 2.483], [12.038999999997031, 4.024, ...   

                                                                          Xaxis  \
ID                                                                                
SN10005   [0.0, 1.0381919191919462, 2.0763838383838924, 3.114575757575839, 4...   
SN10007   [0.0, 1.5136565656566285, 3.027313131313257, 4.540969696969886, 6....   
SN100361  [0.0, 1.3031111111111275, 2.606222222222255, 3.9093333333333824, 5...   
...                                                                         ...   
SN99438   [0.0, 1.7255757575757058, 3.4511515151514116, 5.176727272727117, 6...   
SN99457   [0.0, 1.7667373737373662, 3.5334747474747323, 5.300212121212098, 7...   
SN99946   [0.0, 1.3626161616161374, 2.725232323232275, 4.087848484848412, 5....   

                                                                        desg_GP  \
ID                                                                                
SN10005   ([10.027000000001863, 10.782565656567513, 11.538131313133166, 12.2...   
SN10007   ([0.0, 1.5132626262626456, 3.0265252525252913, 4.539787878787937, ...   
SN1

### Auto ML 5 models 

In [25]:
start_time = time.time()

aml = auto_ML(df, 5, .5)

print("--- %s seconds ---" % (time.time() - start_time))

Checking whether there is an H2O instance running at http://localhost:54323 . connected.


--------------------------  ------------------------------------------------------------------
H2O_cluster_uptime:         13 mins 07 secs
H2O_cluster_timezone:       America/Sao_Paulo
H2O_data_parsing_timezone:  UTC
H2O_cluster_version:        3.30.1.3
H2O_cluster_version_age:    2 months and 24 days
H2O_cluster_name:           H2O_from_python_fmoliveira_wvcquq
H2O_cluster_total_nodes:    1
H2O_cluster_free_memory:    3.871 Gb
H2O_cluster_total_cores:    4
H2O_cluster_allowed_cores:  4
H2O_cluster_status:         locked, healthy
H2O_connection_url:         http://localhost:54323
H2O_connection_proxy:       {"http": null, "https": null}
H2O_internal_security:      False
H2O_API_Extensions:         Amazon S3, XGBoost, Algos, AutoML, Core V3, TargetEncoder, Core V4
Python_version:             3.6.9 final
--------------------------  ------------------------------------------------------------------

Parse progress: |█████████████████████████████████████████████████████████| 100%
AutoML progress: |████████████████████████████████████████████████████████| 100%
Leaderboard:  

model_id mean_residual_deviance rmse mse mae rmsle 
 
 
 StackedEnsemble_AllModels_AutoML_20201222_232526 0.00806606 0.0898112 0.00806606 0.0687872 0.0520735 
 StackedEnsemble_BestOfFamily_AutoML_20201222_232526 0.00832846 0.0912604 0.00832846 0.0703246 0.0528456 
 DRF_1_AutoML_20201222_232526 0.00869452 0.0932444 0.00869452 0.0720666 0.0540014 
 XGBoost_3_AutoML_20201222_232526 0.00912034 0.0955004 0.00912034 0.074035 0.0553491 
 XGBoost_2_AutoML_20201222_232526 0.00966219 0.0982965 0.00966219 0.0745544 0.0569475 
 XGBoost_1_AutoML_20201222_232526 0.0097778 0.0988828 0.0097778 0.0750848 0.057114 
 GLM_1_AutoML_20201222_232526 0.038624 0.19653 0.038624 0.165075 0.118047

 

Leader:  Model Details
H2OStackedEnsembleEstimator :  Stacked Ensemble
Model Key:  StackedEnsemble_AllModels_AutoML_20201222_232526

No model summary for this model

ModelMetricsRegressionGLM: stackedensemble
** Reported on train data. **

MSE: 0.0010907657461668554
RMSE: 0.03302674289370442
MAE: 0.025236380215375002
RMSLE: 0.01911988931095091
R^2: 0.97967080874221
Mean Residual Deviance: 0.0010907657461668554
Null degrees of freedom: 2513
Residual degrees of freedom: 2508
Null deviance: 134.8890396617478
Residual deviance: 2.742185085863474
AIC: -9999.257584319548

ModelMetricsRegressionGLM: stackedensemble
** Reported on cross-validation data. **

MSE: 0.008912626357440483
RMSE: 0.09440670716342395
MAE: 0.07274288565786766
RMSLE: 0.054205061134228696
R^2: 0.8338905613177126
Mean Residual Deviance: 0.008912626357440483
Null degrees of freedom: 2513
Residual degrees of freedom: 2508
Null deviance: 134.9564244916461
Residual deviance: 22.406342662605375
AIC: -4718.37685294119
 

--- 

In [20]:
save_aml_models(aml)

Model StackedEnsemble_AllModels_AutoML_20201222_231222 saved.
Model StackedEnsemble_BestOfFamily_AutoML_20201222_231222 saved.
Model DRF_1_AutoML_20201222_231222 saved.
Model XGBoost_3_AutoML_20201222_231222 saved.
Model XGBoost_2_AutoML_20201222_231222 saved.
Model XGBoost_1_AutoML_20201222_231222 saved.
Model GLM_1_AutoML_20201222_231222 saved.


### Auto ML 50 models 
#### Best model is always StackedEnsemble_AllModels.
#### It was validated with 2543 objects and trained with 2535
##### Values set by ` h2o.H2OFrame( df.loc[:,'f1':'REDSHIFT_SPEC'] ).split_frame(ratios=[.5]) `

In [26]:
start_time = time.time()

aml = auto_ML(df, 50, .5)

print("--- %s seconds ---" % (time.time() - start_time))

Checking whether there is an H2O instance running at http://localhost:54323 . connected.


--------------------------  ------------------------------------------------------------------
H2O_cluster_uptime:         14 mins 44 secs
H2O_cluster_timezone:       America/Sao_Paulo
H2O_data_parsing_timezone:  UTC
H2O_cluster_version:        3.30.1.3
H2O_cluster_version_age:    2 months and 24 days
H2O_cluster_name:           H2O_from_python_fmoliveira_wvcquq
H2O_cluster_total_nodes:    1
H2O_cluster_free_memory:    3.861 Gb
H2O_cluster_total_cores:    4
H2O_cluster_allowed_cores:  4
H2O_cluster_status:         locked, healthy
H2O_connection_url:         http://localhost:54323
H2O_connection_proxy:       {"http": null, "https": null}
H2O_internal_security:      False
H2O_API_Extensions:         Amazon S3, XGBoost, Algos, AutoML, Core V3, TargetEncoder, Core V4
Python_version:             3.6.9 final
--------------------------  ------------------------------------------------------------------

Parse progress: |█████████████████████████████████████████████████████████| 100%
AutoML progress: |████████████████████████████████████████████████████████| 100%
Leaderboard:  

model_id mean_residual_deviance rmse mse mae rmsle 
 
 
 StackedEnsemble_AllModels_AutoML_20201222_232703 0.00699907 0.0836604 0.00699907 0.0646625 0.0481486 
 StackedEnsemble_BestOfFamily_AutoML_20201222_232703 0.00722633 0.0850078 0.00722633 0.065425 0.0486164 
 GBM_4_AutoML_20201222_232703 0.00790305 0.0888991 0.00790305 0.0677892 0.0509523 
 GBM_3_AutoML_20201222_232703 0.00800494 0.0894703 0.00800494 0.0688634 0.0511717 
 GBM_grid__1_AutoML_20201222_232703_model_2 0.00806419 0.0898008 0.00806419 0.0690973 0.0515418 
 GBM_1_AutoML_20201222_232703 0.00825027 0.090831 0.00825027 0.0705596 0.0521052 
 GBM_2_AutoML_20201222_232703 0.00835021 0.0913795 0.00835021 0.0706635 0.0523642 
 XRT_1_AutoML_20201222_232703 0.0084333 0.091833 0.0084333 0.0715695 0.052871 
 DRF_1_AutoML_20201222_232703 0.00847876 0.0920802 0.00847876 0.071747 0.052915 
 XGBoost_grid__1_AutoML_20201222_232703_model_14 0.00848845 0.0921328 0.00848845 0.0710784 0.053004 
 XGBoost_grid__1_AutoML_20201222_232703_model_3 0.00855088 0.092471 0.00855088 0.0715582 0.0533484 
 GBM_grid__1_AutoML_20201222_232703_model_3 0.00858626 0.0926621 0.00858626 0.071452 0.0530792 
 XGBoost_grid__1_AutoML_20201222_232703_model_16 0.0086277 0.0928854 0.0086277 0.0726957 0.0535326 
 XGBoost_grid__1_AutoML_20201222_232703_model_5 0.00868511 0.0931939 0.00868511 0.0722951 0.0535316 
 XGBoost_3_AutoML_20201222_232703 0.00872583 0.0934122 0.00872583 0.0721619 0.0537919 
 XGBoost_grid__1_AutoML_20201222_232703_model_1 0.00873557 0.0934643 0.00873557 0.0727716 0.0540302 
 GBM_5_AutoML_20201222_232703 0.00878378 0.0937218 0.00878378 0.0731972 0.0544365 
 GBM_grid__1_AutoML_20201222_232703_model_5 0.00879531 0.0937833 0.00879531 0.0721784 0.0540321 
 XGBoost_grid__1_AutoML_20201222_232703_model_12 0.0088557 0.0941047 0.0088557 0.0727826 0.0542569 
 GBM_grid__1_AutoML_20201222_232703_model_7 0.00885971 0.0941261 0.00885971 0.0740815 0.0541136 
 GBM_grid__1_AutoML_20201222_232703_model_10 0.00888492 0.0942599 0.00888492 0.0741071 0.0543649 
 XGBoost_grid__1_AutoML_20201222_232703_model_2 0.00891659 0.0944277 0.00891659 0.0729633 0.0546101 
 XGBoost_grid__1_AutoML_20201222_232703_model_10 0.00894169 0.0945605 0.00894169 0.0736765 0.0549373 
 GBM_grid__1_AutoML_20201222_232703_model_8 0.0090335 0.0950447 0.0090335 0.0740585 0.0545646 
 XGBoost_grid__1_AutoML_20201222_232703_model_17 0.00917623 0.0957927 0.00917623 0.0747268 0.0550153 
 GBM_grid__1_AutoML_20201222_232703_model_6 0.0091881 0.0958546 0.0091881 0.0751187 0.0557321 
 DeepLearning_grid__3_AutoML_20201222_232703_model_3 0.0092204 0.0960229 0.0092204 0.0773092 0.0582063 
 GBM_grid__1_AutoML_20201222_232703_model_9 0.00925907 0.0962241 0.00925907 0.0759884 0.0560609 
 XGBoost_grid__1_AutoML_20201222_232703_model_7 0.00927406 0.0963019 0.00927406 0.0749735 0.0555118 
 XGBoost_grid__1_AutoML_20201222_232703_model_13 0.00928985 0.0963839 0.00928985 0.0744822 0.0554605 
 XGBoost_grid__1_AutoML_20201222_232703_model_11 0.00931169 0.0964971 0.00931169 0.0738388 0.0553173 
 XGBoost_grid__1_AutoML_20201222_232703_model_6 0.00931599 0.0965194 0.00931599 0.0742763 0.0553838 
 XGBoost_grid__1_AutoML_20201222_232703_model_18 0.00952389 0.0975904 0.00952389 0.0748036 0.0557831 
 XGBoost_grid__1_AutoML_20201222_232703_model_15 0.00955703 0.0977601 0.00955703 0.0749438 0.0557859 
 XGBoost_grid__1_AutoML_20201222_232703_model_4 0.00957621 0.0978581 0.00957621 0.0751835 0.0560834 
 GBM_grid__1_AutoML_20201222_232703_model_1 0.00959456 0.0979518 0.00959456 0.0769211 0.0566197 
 XGBoost_grid__1_AutoML_20201222_232703_model_9 0.00969585 0.0984675 0.00969585 0.0767853 0.056628 
 XGBoost_2_AutoML_20201222_232703 0.00970974 0.098538 0.00970974 0.0762796 0.0564414 
 XGBoost_grid__1_AutoML_20201222_232703_model_8 0.00995174 0.0997584 0.00995174 0.0765631 0.0571073 
 XGBoost_1_AutoML_20201222_232703 0.00998792 0.0999396 0.00998792 0.0772476 0.0574264 
 GBM_grid__1_AutoML_20201222_232703_model_4 0.00999746 0.0999873 0.00999746 0.0788049 0.0582918 
 DeepLearnin

 

Leader:  Model Details
H2OStackedEnsembleEstimator :  Stacked Ensemble
Model Key:  StackedEnsemble_AllModels_AutoML_20201222_232703

No model summary for this model

ModelMetricsRegressionGLM: stackedensemble
** Reported on train data. **

MSE: 0.0028240076947590224
RMSE: 0.05314139342131539
MAE: 0.03964403266560037
RMSLE: 0.030559807834808828
R^2: 0.9464039800321321
Mean Residual Deviance: 0.0028240076947590224
Null degrees of freedom: 2488
Residual degrees of freedom: 2474
Null deviance: 131.14696121967106
Residual deviance: 7.0289551522552065
AIC: -7513.953987581311

ModelMetricsRegressionGLM: stackedensemble
** Reported on cross-validation data. **

MSE: 0.008238564930985948
RMSE: 0.09076654081205227
MAE: 0.06912853317652369
RMSLE: 0.05221327136720305
R^2: 0.8436426744278345
Mean Residual Deviance: 0.008238564930985948
Null degrees of freedom: 2488
Residual degrees of freedom: 2473
Null deviance: 131.5502154261692
Residual deviance: 20.505788113224025
AIC: -4847.058534229911
 



In [28]:
print(f'{5625/3600} hours training.')

1.5625 hours training.


In [27]:
save_aml_models(aml)

Model StackedEnsemble_AllModels_AutoML_20201222_232703 saved.
Model StackedEnsemble_BestOfFamily_AutoML_20201222_232703 saved.
Model GBM_4_AutoML_20201222_232703 saved.
Model GBM_3_AutoML_20201222_232703 saved.
Model GBM_grid__1_AutoML_20201222_232703_model_2 saved.
Model GBM_1_AutoML_20201222_232703 saved.
Model GBM_2_AutoML_20201222_232703 saved.
Model XRT_1_AutoML_20201222_232703 saved.
Model DRF_1_AutoML_20201222_232703 saved.
Model XGBoost_grid__1_AutoML_20201222_232703_model_14 saved.
Model XGBoost_grid__1_AutoML_20201222_232703_model_3 saved.
Model GBM_grid__1_AutoML_20201222_232703_model_3 saved.
Model XGBoost_grid__1_AutoML_20201222_232703_model_16 saved.
Model XGBoost_grid__1_AutoML_20201222_232703_model_5 saved.
Model XGBoost_3_AutoML_20201222_232703 saved.
Model XGBoost_grid__1_AutoML_20201222_232703_model_1 saved.
Model GBM_5_AutoML_20201222_232703 saved.
Model GBM_grid__1_AutoML_20201222_232703_model_5 saved.
Model XGBoost_grid__1_AutoML_20201222_232703_model_12 saved.
Mo

### Making predictions.

In [71]:
best_model = h2o.import_mojo('../../models/mojo_50_ensemble/StackedEnsemble_AllModels_AutoML_20201222_232703.zip')
preds, len_train, len_validation = gen_predictions(df, best_model)

generic Model Build progress: |███████████████████████████████████████████| 100%
Model Details
H2OGenericEstimator :  Import MOJO Model
Model Key:  Generic_model_python_1608689537668_7

No model summary for this model

ModelMetricsRegressionGeneric: generic
** Reported on train data. **

MSE: 0.0028240076947590224
RMSE: 0.05314139342131539
MAE: 0.03964403266560037
RMSLE: 0.030559807834808828
Mean Residual Deviance: 0.0028240076947590224

ModelMetricsRegressionGeneric: generic
** Reported on cross-validation data. **

MSE: 0.008238564930985948
RMSE: 0.09076654081205227
MAE: 0.06912853317652369
RMSLE: 0.05221327136720305
Mean Residual Deviance: 0.008238564930985948

Parse progress: |█████████████████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%
generic prediction progress: |████████████████████████████████████████████| 100%

 2550 train/test objects  
 and 2528 validation objects


### Manual RMSE verification

In [73]:
pd_preds = preds.as_data_frame()['predict'].tolist()
true_y = list(df.iloc[-len_validation:].loc[:,'REDSHIFT_SPEC'])
RMSE = (mean_squared_error(list(df.iloc[-len_validation:].loc[:,'REDSHIFT_SPEC']), pd_preds))**(1/2)
RMSE

0.07027877615784286

### ANS: Values don't match exactly with leaderboard RMSE Score ( 0.0836604	). However SckiKit external score reachs an even lower value. Approved and Verified.